### PR TITLE
Adding revealed messages to challenge hash

### LINKED ISSUE #74: Adding the revealed messages to the chalenge to avoid forgery
The following is a method that could possibly allow the prover to reveal messages for which they don't have a signature, taking advantage of the Fiat–Shamir challenge not containing the revealed messages. Let the prover with a signature `(A, e, s)` on messages `msg_1`, `msg_2` and `msg_3`, creating an spk revealing only `msg_1`. Going through the spkGen process the prover will calculate
```
C2_prover = D*(-r3~) +H0 * (s~) + H_2 * m~_2 + H_3 * m~_3
```
The proof value will be,
```
spk = (Abar, A’, D, c, e^, r2^, r3^, s^, m^_2, m^_3)
```
From the verification process we know that if `T = P1 + H_1 * msg_1` then 

```
#EQ1: T*c + D * (-r3^) + H0*s^ + H_2 * m^_2 + H_3 * m^_3 = C2_prover
```

Let the prover now calculate the following,
1.	`c’  = c ^ -1 mod q`
2.	`msg_adv  = m^_2 * c'`

and instead of revealing spk, they send `spk2`,
```
spk2 = (Abar, A’, D, c, e^, r2^, r3^, s^, m^_3)
```
with revealed messages `msg_1` and `msg_adv`. 

The verifier receiving `spk2` and `msg_1`, `msg_adv` will calculate `C1_verifier`, `C2_verifier` and check the pairing equality. Note that since `Abar, A’, D, c, e^, r2^` remain the same between `spk` and `spk2`, `C1_prover = C1_verifier` and `e(A', PK) = e(Abar, P2)`. To calculate `C2_verifier` the verifier will first calculate `T2` as,

```
T2 = P1 + H_1 * msg_1 + H_2 * msg_adv = P1 + H_1 * msg_1 + H_2 * (m^_2 * c')
```
From that the verifier will calculate,
```
C2_verifier = T2*c + D * (-r3^) + H0*s^ + H_3 * m^_3 =
   = D * (-r3^) + H0*s^ + (P1 + H_1 * msg_1) * c +  H_2 * (m^_2 * c' * c)  H_3 * m^_3 =  
   = D * (-r3^) + H0*s^ + T * c +  H_2 * m^_2 +  H_3 * m^_3  = C2_prover
```
The last equality follows from #EQ1. As a result, the verification will succeed for `msg_1` and `msg_adv`.

Obviously, this should not be possible, since this proves knowledge of a signature that the prover doesn't have (which contains `msg_adv`). If I'm correct on the above, most likely the only issue is with the Fiat–Shamir heuristic. Note that if the protocol was interactive, that method will not work, since the prover would have to "commit" to the messages prior of getting the challenge `c`. Hence we should also pass the revealed messages to the hash of the challenge during spkGen and spkVerify, to "commit" them before the challenge is created.

### PR DESCRIPTION
**Pull request by BasileiosKal**
Closes #74 

Adding the revealed messages as well as the signature domain value to the challenge hash in ProofGen and ProofVerify.

### CODE REVIEW

**Review by tplooker — APPROVED** [review]
LGTM

File: draft-bbs-signatures.md [hunk 1]
```diff
@@ -534,45 +534,47 @@ Procedure:
...
+17. s' = s + r2 * r3
 
-17. C1 = A' * e~ + H_s * r2~
+18. C1 = A' * e~ + H_s * r2~
 
-18. C2 = D * (-r3~) + H_s * s~ + H_j1 * m~_j1 + ... + H_jU * m~_jU
+19. C2 = D * (-r3~) + H_s * s~ + H_j1 * m~_j1 + ... + H_jU * m~_jU
 
-19. c = HASH(PK || Abar || A' || D || C1 || C2 || ph)
+20. c = HASH(PK || Abar || A' || D || C1 || C2 || domain || revealed_msgs || ph)
```
Comment (tplooker) [inline draft-bbs-signatures.md:563]: One minor question do we want to include `R` alongside the actual revealed messages like we include L alongside the generators in the domain?

**Comment by BasileiosKal** [comment]
Couple of updates, 
1. Per @tplooker suggestion the R value (number of revealed messages) is added to the challenge hash.
2. Removed the PK from the challenge hash since it is already part of the domain value.

File: draft-bbs-signatures.md (hunk 1 above)
Comment (BasileiosKal) [reply draft-bbs-signatures.md]: Done. Thanks!

File: draft-bbs-signatures.md [hunk 2]
```diff
@@ -534,45 +534,47 @@ Procedure:
 
 6. generators =  (H_s || H_d || H_1 || ... || H_L)
 
-7. domain = OS2IP(HASH(PK || L || generators || Ciphersuite_ID || header)) mod q
+7. revealed_msgs = (msg_i1 || msg_i2 || ... || msg_iR)
```
Comment (tplooker) [inline draft-bbs-signatures.md:537]: On further thought with ProofGen, where is (msg_i1, .., msg_iR) defined?

Comment (tplooker) [reply draft-bbs-signatures.md:537]: We need to add a step to this procedure that computes msg_i1 ... msg_iR from msg and RevealedIndexes

File: draft-bbs-signatures.md [hunk 3]
```diff
@@ -505,6 +505,7 @@ Inputs:
 - msg_1,..., msg_L, octet strings. Messages in input to Sign.
 - H_1,..., H_L, points of G1. The generators in input to Sign.
 - RevealedIndexes, vector of unsigned integers. Indexes of revealed messages.
+                   RevealedIndexes MUST be ordered from the lowest to the highest value.
```
Comment (tplooker) [inline draft-bbs-signatures.md:508]:
```suggestion
                   This vector MUST be ordered from the lowest to the highest value.
```

File: draft-bbs-signatures.md [hunk 4]
```diff
@@ -591,6 +592,7 @@ Inputs:
 - msg_i1,..., msg_iR, octet strings. The revealed messages in input to ProofGen.
 - H_1,..., H_L, points of G1. The generators in input to Sign.
 - RevealedIndexes, vector of unsigned integers. Indexes of revealed messages.
+                   RevealedIndexes MUST be ordered from the lowest to the highest value.
```
Comment (tplooker) [inline draft-bbs-signatures.md:595]:
```suggestion
                   This vector MUST be ordered from the lowest to the highest value.
```

**Review by tplooker — CHANGES_REQUESTED** [review]
Have updated with a comment about a way to make the procedure clearer

File: draft-bbs-signatures.md [hunk 5]
```diff
@@ -534,45 +535,47 @@ Procedure:
...
+17. s' = s + r2 * r3
 
-17. C1 = A' * e~ + H_s * r2~
+18. C1 = A' * e~ + H_s * r2~
 
-18. C2 = D * (-r3~) + H_s * s~ + H_j1 * m~_j1 + ... + H_jU * m~_jU
+19. C2 = D * (-r3~) + H_s * s~ + H_j1 * m~_j1 + ... + H_jU * m~_jU
 
-19. c = HASH(PK || Abar || A' || D || C1 || C2 || ph)
+20. c = HASH(Abar || A' || D || C1 || C2 || R || revealed_msgs || domain || ph)
```
Comment (mikelodder7) [inline draft-bbs-signatures.md:564]: We should add the messages too and not just the indices. We need them both. If that’s already the case it’s not clear to me how that is so given revealed_msgs appears to just include indices

Comment (BasileiosKal) [reply draft-bbs-signatures.md:564]: Thank you! Currently the PR only includes the revealed messages (note that `revealed_msgs =  (msg_i1 || msg_i2 || ... || msg_iR)` where `msg_i1, ..., msg_iR` the revealed messages in the current notation). Including the revealed indices also makes sense to me but there are proposals to remove the indices notion altogether (see for example issue #117 for some context). For that reason, I made the PR a draft for now, until a proper notation is decided.

**Comment by tplooker** [comment]
Will be addressed after merging of #138

**Comment by BasileiosKal** [comment]
Replaced by #188. Closing.

**Pull request closed without merging**